File: bindings/android/src/main/kotlin/com/microsoft/ai/foundry/local/mobile/Session.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.ai.foundry.local.mobile

import com.microsoft.ai.foundry.local.mobile.internal.JobBridge
import com.microsoft.ai.foundry.local.mobile.internal.JsonCodec
import com.microsoft.ai.foundry.local.mobile.internal.NativeBridge
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Common base for all inference sessions.
 *
 * A session owns a KV cache and a small amount of runtime state on top of the
 * loaded model. Release with [close] when done; leaving a session alive
 * indefinitely holds memory that a background trim would otherwise reclaim.
 */
public abstract class Session internal constructor(
    protected val model: Model,
    protected val handle: Long,
) : AutoCloseable {

    private val closed = AtomicBoolean(false)

    /**
     * Update mutable session settings. Same schema as the session's
     * constructor options.
     */
    protected fun setNativeOptions(optionsJson: String) {
        NativeBridge.sessionSetOptions(requireHandle(), optionsJson)
    }

    /** Number of completed turns in the conversation. */
    public val turnCount: Long get() = NativeBridge.sessionGetTurnCount(requireHandle())

    /** Drop the last [count] turns. */
    public fun undoTurns(count: Long) {
        NativeBridge.sessionUndoTurns(requireHandle(), count)
    }

    /** Discard all conversation history, keeping the session and its options. */
    public fun clearHistory() {
        NativeBridge.sessionClearHistory(requireHandle())
    }

    /**
     * Serialize the session so it can be restored after the process is
     * killed. Returns the JSON payload.
     */
    public fun exportHistory(): String = NativeBridge.sessionExportHistoryJson(requireHandle())

    /** Restore history previously produced by [exportHistory]. */
    public fun restoreHistory(historyJson: String) {
        NativeBridge.sessionRestoreHistoryJson(requireHandle(), historyJson)
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return
        try { NativeBridge.sessionRelease(handle) } catch (_: Throwable) {}
    }

    @Suppress("removal", "deprecation")
    protected fun finalize() {
        if (!closed.get()) close()
    }

    protected fun requireHandle(): Long {
        check(!closed.get()) { "Session has been closed" }
        return handle
    }
}

// -----------------------------------------------------------------------------
// Chat
// -----------------------------------------------------------------------------

public class ChatSession internal constructor(model: Model, options: ChatOptions) :
    Session(model, createNative(model, options)) {

    /** Update the session's chat options. */
    public fun updateOptions(options: ChatOptions) {
        setNativeOptions(encodeOptions("chat", options))
    }

    /**
     * Run a chat completion in streaming mode. The [prompt] is added as a
     * single user turn to the session history.
     *
     * Emits **only** text fragments — tool calls, usage accounting and the
     * completion event are dropped. That matches what a chat UI typically
     * wants; a caller that needs those events (a tool-calling agent, a
     * token-cost meter, a UI that lights up on `FinishReason.LENGTH`) must
     * build a [ChatRequest] and use [completeAllDeltas] instead.
     */
    public fun completeStreaming(prompt: String): Flow<Delta.Text> = completeStreaming(userTurn(prompt))

    /**
     * Streaming completion with a raw OpenAI-shaped [ChatRequest].
     *
     * Emits **only** text fragments — tool calls, usage accounting and the
     * completion event are dropped. See [completeAllDeltas] for the
     * everything-flavoured overload, and see [complete] for a non-streaming
     * variant that returns tool calls and usage on a single result object.
     */
    public fun completeStreaming(request: ChatRequest): Flow<Delta.Text> =
        JobBridge.stream { corr ->
            NativeBridge.sessionCompleteAsync(requireHandle(), request.toJson(), true, corr)
        }.mapNotNull { it as? Delta.Text }

    /**
     * Streaming completion emitting **every** delta kind — text fragments,
     * reasoning fragments, tool calls, usage accounting and the terminal
     * `Delta.Completed`.
     *
     * Reach for this when the app needs tool call events or per-turn token
     * counts. Chat UIs that only need to paint text into a bubble should
     * prefer [completeStreaming], whose narrower return type
     * ([kotlinx.coroutines.flow.Flow] of [Delta.Text]) is easier to consume
     * — the type says "you get text or nothing", so callers do not have to
     * `is` check every element.
     */
    public fun completeAllDeltas(request: ChatRequest): Flow<Delta> =
        JobBridge.stream { corr ->
            NativeBridge.sessionCompleteAsync(requireHandle(), request.toJson(), true, corr)
        }

    /** Non-streaming completion. Returns the whole assistant text. */
    public suspend fun complete(prompt: String): CompleteResult = complete(userTurn(prompt))

    public suspend fun complete(request: ChatRequest): CompleteResult {
        val json = JobBridge.awaitResult { corr ->
            NativeBridge.sessionCompleteAsync(requireHandle(), request.toJson(), false, corr)
        } ?: return CompleteResult(text = "", finishReason = FinishReason.NONE)
        val obj = JsonCodec.parseObject(json)
        val text = obj["text"]?.jsonPrimitive?.content ?: ""
        val finishReason = FinishReason.fromString(obj["finish_reason"]?.jsonPrimitive?.content)
        // The ABI treats absent-vs-empty for `tool_calls` and `usage` as
        // meaningful: absent = the runtime had nothing to report, empty =
        // could not happen because the empty case is reported as absent.
        // Preserve the distinction by using null for "absent" instead of
        // collapsing to an empty list.
        val toolCalls = (obj["tool_calls"] as? JsonArray)?.mapNotNull { el ->
            (el as? JsonObject)?.let { tc ->
                Delta.ToolCall(
                    callId = tc["call_id"]?.jsonPrimitive?.content ?: return@mapNotNull null,
                    name = tc["name"]?.jsonPrimitive?.content ?: return@mapNotNull null,
                    // Wire form: arguments is a JSON *string* (double-encoded)
                    // because the model may emit something that does not match
                    // the tool schema, and it's the app's job to decide.
                    argumentsJson = tc["arguments"]?.jsonPrimitive?.content ?: "{}",
                )
            }
        }
        val usage = (obj["usage"] as? JsonObject)?.let { u ->
            Delta.Usage(
                promptTokens = u["prompt_tokens"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                completionTokens = u["completion_tokens"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
            )
        }
        return CompleteResult(text, finishReason, toolCalls, usage, json)
    }

    /**
     * Continue a turn by delivering results for tools the model asked to call.
     * [results] pairs each `call_id` with the JSON-encoded return value.
     */
    public fun submitToolResults(results: List<ToolResult>): Flow<Delta> {
        val payload = buildJsonArray {
            results.forEach { r ->
                addJsonObject {
                    put("call_id", r.callId)
                    // The ABI takes `result` as a stringified JSON, so we pass
                    // the caller's raw JSON string through unchanged.
                    put("result", r.resultJson)
                }
            }
        }.toString()
        return JobBridge.stream { corr ->
            NativeBridge.sessionSubmitToolResultsAsync(requireHandle(), payload, true, corr)
        }
    }

    public companion object {
        private fun createNative(model: Model, options: ChatOptions): Long =
            NativeBridge.sessionCreate(model.nativeHandle(), encodeOptions("chat", options))

        private fun encodeOptions(type: String, options: ChatOptions): String = buildJsonObject {
            put("type", type)
            options.systemPrompt?.let { put("system_prompt", it) }
            options.temperature?.let { put("temperature", it) }
            options.topP?.let { put("top_p", it) }
            options.topK?.let { put("top_k", it) }
            options.maxOutputTokens?.let { put("max_output_tokens", it) }
            options.seed?.let { put("seed", it) }
            put("keep_history", options.keepHistory)
        }.toString()

        private fun userTurn(text: String): ChatRequest = ChatRequest(
            messages = listOf(ChatMessage(role = "user", content = text)),
        )
    }
}

// -----------------------------------------------------------------------------
// Audio
// -----------------------------------------------------------------------------

public class AudioSession internal constructor(model: Model, options: AudioOptions) :
    Session(model, createNative(model, options)) {

    /** Transcribe a file already on disk, or a base64-encoded blob. */
    public fun transcribeStreaming(request: TranscribeRequest): Flow<Delta> =
        JobBridge.stream { corr ->
            NativeBridge.sessionTranscribeAsync(requireHandle(), request.toJson(), true, corr)
        }

    /** Transcribe [request] and return the full result (text, language, segments). */
    public suspend fun transcribe(request: TranscribeRequest): TranscribeResult {
        val json = JobBridge.awaitResult { corr ->
            NativeBridge.sessionTranscribeAsync(requireHandle(), request.toJson(), false, corr)
        } ?: return TranscribeResult(text = "", language = null, segments = emptyList())
        val obj = JsonCodec.parseObject(json)
        val text = obj["text"]?.jsonPrimitive?.content ?: ""
        val language = obj["language"]?.jsonPrimitive?.content
        val segments = (obj["segments"] as? JsonArray)?.mapNotNull { el ->
            (el as? JsonObject)?.let { seg ->
                TranscribeSegment(
                    text = seg["text"]?.jsonPrimitive?.content ?: "",
                    startTimeMs = seg["start_time_ms"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                    endTimeMs = seg["end_time_ms"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                    language = seg["language"]?.jsonPrimitive?.content,
                )
            }
        } ?: emptyList()
        return TranscribeResult(text, language, segments)
    }

    /**
     * Push a chunk of PCM audio into a live transcription session.
     *
     * The corresponding streaming call — `transcribeStreaming` — must have
     * been started first, with `{ "streaming": true }` in the request. Partial
     * and final segments arrive on that flow.
     */
    public fun pushAudio(pcm: ByteArray, sampleRate: Int, channels: Int, isFinal: Boolean = false) {
        NativeBridge.sessionPushAudio(requireHandle(), pcm, sampleRate, channels, isFinal)
    }

    public companion object {
        private fun createNative(model: Model, options: AudioOptions): Long =
            NativeBridge.sessionCreate(
                model.nativeHandle(),
                buildJsonObject {
                    put("type", "audio")
                    options.language?.let { put("language", it) }
                    options.maxOutputTokens?.let { put("max_output_tokens", it) }
                }.toString(),
            )
    }
}

// -----------------------------------------------------------------------------
// Embedding
// -----------------------------------------------------------------------------

public class EmbeddingSession internal constructor(model: Model, options: EmbeddingOptions) :
    Session(model, createNative(model, options)) {

    /** Compute embeddings for [inputs]. */
    public suspend fun embed(inputs: List<String>): EmbeddingResult {
        val request = buildJsonObject {
            putJsonArray("inputs") { inputs.forEach { add(it) } }
        }.toString()
        val json = JobBridge.awaitResult { corr ->
            NativeBridge.sessionEmbedAsync(requireHandle(), request, corr)
        } ?: return EmbeddingResult(emptyList(), 0)
        val obj = JsonCodec.parseObject(json)
        val dims = obj["dimensions"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0
        val vectors = (obj["embeddings"] as? JsonArray)?.map { row ->
            (row as? JsonArray)?.map { (it as JsonPrimitive).content.toFloatOrNull() ?: 0f } ?: emptyList()
        } ?: emptyList()
        return EmbeddingResult(vectors, dims)
    }

    public companion object {
        private fun createNative(model: Model, options: EmbeddingOptions): Long =
            NativeBridge.sessionCreate(
                model.nativeHandle(),
                buildJsonObject {
                    put("type", "embedding")
                }.toString(),
            )
    }
}

// -----------------------------------------------------------------------------
// Request / result types
// -----------------------------------------------------------------------------

public data class ChatRequest(
    val messages: List<ChatMessage>,
    val tools: List<Tool> = emptyList(),
    val toolChoice: String? = null,
    val temperature: Float? = null,
    val maxOutputTokens: Int? = null,
) {
    internal fun toJson(): String = buildJsonObject {
        putJsonArray("messages") {
            messages.forEach { m ->
                addJsonObject {
                    put("role", m.role)
                    when (val content = m.content) {
                        is ChatContent.Text -> put("content", content.text)
                        is ChatContent.Parts -> putJsonArray("content") {
                            content.parts.forEach { p -> add(p.toJson()) }
                        }
                    }
                }
            }
        }
        if (tools.isNotEmpty()) {
            putJsonArray("tools") {
                tools.forEach { t ->
                    addJsonObject {
                        put("name", t.name)
                        put("description", t.description)
                        // parametersJson is a raw JSON Schema document — embed
                        // it verbatim rather than as a quoted string.
                        put(
                            "parameters",
                            kotlinx.serialization.json.Json.parseToJsonElement(t.parametersJson),
                        )
                    }
                }
            }
        }
        toolChoice?.let { put("tool_choice", it) }
        temperature?.let { put("temperature", it) }
        maxOutputTokens?.let { put("max_output_tokens", it) }
    }.toString()
}

public data class ChatMessage(
    val role: String,
    val content: ChatContent,
) {
    public constructor(role: String, content: String) : this(role, ChatContent.Text(content))
}

public sealed class ChatContent {
    public data class Text(val text: String) : ChatContent()
    public data class Parts(val parts: List<ChatContentPart>) : ChatContent()
}

public sealed class ChatContentPart {
    public data class Text(val text: String) : ChatContentPart()
    public data class Image(val path: String? = null, val dataBase64: String? = null) : ChatContentPart()
    public data class Audio(val path: String? = null, val dataBase64: String? = null) : ChatContentPart()

    internal fun toJson(): JsonElement = when (this) {
        is Text -> buildJsonObject {
            put("type", "text")
            put("text", text)
        }
        is Image -> buildJsonObject {
            put("type", "image")
            path?.let { put("path", it) }
            dataBase64?.let { put("data_base64", it) }
        }
        is Audio -> buildJsonObject {
            put("type", "audio")
            path?.let { put("path", it) }
            dataBase64?.let { put("data_base64", it) }
        }
    }
}

public data class Tool(
    val name: String,
    val description: String,
    /**
     * The JSON Schema for this tool's parameters, as a raw JSON string. Passing
     * a string keeps the type surface small; typical apps store the schema as
     * a resource asset and read it into a string.
     */
    val parametersJson: String,
)

public data class ToolResult(
    val callId: String,
    val resultJson: String,
)

public sealed class TranscribeRequest {
    public data class File(val path: String, val language: String? = null, val translate: Boolean = false) :
        TranscribeRequest()

    public data class InMemory(
        val dataBase64: String,
        val format: String = "pcm",
        val sampleRate: Int = 16000,
        val channels: Int = 1,
    ) : TranscribeRequest()

    /**
     * Start a streaming transcription that receives audio via
     * [AudioSession.pushAudio].
     */
    public data class Streaming(val language: String? = null) : TranscribeRequest()

    internal fun toJson(): String = when (this) {
        is File -> buildJsonObject {
            put("path", path)
            language?.let { put("language", it) }
            put("translate", translate)
        }.toString()
        is InMemory -> buildJsonObject {
            put("data_base64", dataBase64)
            put("format", format)
            put("sample_rate", sampleRate)
            put("channels", channels)
        }.toString()
        is Streaming -> buildJsonObject {
            put("streaming", true)
            language?.let { put("language", it) }
        }.toString()
    }
}

public data class CompleteResult(
    val text: String,
    val finishReason: FinishReason,
    /**
     * Tool calls the model wants executed, or `null` when the model made no
     * tool calls. Nullable — not defaulted to an empty list — to preserve the
     * absent-vs-empty distinction the ABI carries. In practice the empty case
     * does not appear on the wire, but a caller writing exhaustive tool
     * dispatch logic should still handle `null` explicitly.
     */
    val toolCalls: List<Delta.ToolCall>? = null,
    /** Token accounting, or `null` if the runtime did not report it. */
    val usage: Delta.Usage? = null,
    /** The raw completion JSON, exposed for callers who need unmodelled fields. */
    val rawJson: String? = null,
)

public data class EmbeddingResult(
    val embeddings: List<List<Float>>,
    val dimensions: Int,
)

public data class TranscribeResult(
    val text: String,
    /** Detected language BCP-47 tag, or null when the model reports none. */
    val language: String?,
    /** Aligned segments; empty when the runtime does not report timestamps. */
    val segments: List<TranscribeSegment>,
)

public data class TranscribeSegment(
    val text: String,
    val startTimeMs: Long,
    val endTimeMs: Long,
    val language: String?,
)

File: bindings/react-native/src/Session.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import NativeFoundryLocal from './NativeFoundryLocal';
import { FoundryLocalError } from './errors';
import { makeDeltaStream, nextSubscriptionId } from './stream';
import type {
  AudioOptions,
  ChatOptions,
  ChatRequest,
  CompleteResult,
  Delta,
  EmbeddingOptions,
  EmbeddingResult,
  FinishReason,
  ToolCall,
  ToolResult,
  TranscribeRequest,
  TranscribeResult,
  UsageStats,
} from './types';
import type { Model } from './Model';

const FINISH_REASONS: readonly FinishReason[] = [
  'none', 'stop', 'length', 'tool_calls', 'cancelled', 'error',
];

function parseFinishReason(s: string | undefined | null): FinishReason {
  if (s == null || s === '') return 'none';
  return FINISH_REASONS.includes(s as FinishReason) ? (s as FinishReason) : 'unknown';
}

function toNumber(value: unknown): number {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string') {
    const n = Number(value);
    if (Number.isFinite(n)) return n;
  }
  return 0;
}

/**
 * Common base for every session type. Owns a native handle and unregisters
 * it on `close()`. `Session` is disposable — a released session invalidates
 * every subsequent call.
 */
export abstract class Session {
  protected readonly handle: number;
  protected closed = false;

  protected constructor(handle: number) {
    this.handle = handle;
  }

  /** Number of completed turns in the conversation. */
  public get turnCount(): number {
    this.checkOpen();
    return NativeFoundryLocal.sessionGetTurnCount(this.handle);
  }

  public undoTurns(count: number): void {
    this.checkOpen();
    NativeFoundryLocal.sessionUndoTurns(this.handle, count);
  }

  public clearHistory(): void {
    this.checkOpen();
    NativeFoundryLocal.sessionClearHistory(this.handle);
  }

  /** JSON-encoded conversation history the app can persist and re-load. */
  public exportHistory(): string {
    this.checkOpen();
    return NativeFoundryLocal.sessionExportHistory(this.handle);
  }

  public restoreHistory(historyJson: string): void {
    this.checkOpen();
    NativeFoundryLocal.sessionRestoreHistory(this.handle, historyJson);
  }

  public close(): void {
    if (this.closed) return;
    this.closed = true;
    try { NativeFoundryLocal.sessionRelease(this.handle); }
    catch { /* releasing an already-freed session is a no-op */ }
  }

  protected checkOpen(): void {
    if (this.closed) {
      throw new FoundryLocalError(4, 'Session has been closed', null, 'invalidState');
    }
  }
}

// -----------------------------------------------------------------------------
// Chat
// -----------------------------------------------------------------------------

/**
 * A chat / completion session. Instances hold a KV cache and a small amount
 * of state on top of the loaded model.
 */
export class ChatSession extends Session {
  /** @internal */
  public static create(model: Model, options: ChatOptions): ChatSession {
    const handle = NativeFoundryLocal.sessionCreate(
      model.nativeHandle,
      JSON.stringify({ type: 'chat', ...encodeChatOptions(options) }),
    );
    return new ChatSession(handle);
  }

  private constructor(handle: number) {
    super(handle);
  }

  public updateOptions(options: ChatOptions): void {
    this.checkOpen();
    NativeFoundryLocal.sessionSetOptions(
      this.handle,
      JSON.stringify({ type: 'chat', ...encodeChatOptions(options) }),
    );
  }

  /**
   * Streaming completion. Iterate with `for await`. Breaking out of the
   * loop cancels the underlying job.
   *
   * The stream only yields text deltas by default. Use
   * {@link completeAllDeltas} when the caller needs tool-call events or
   * mid-generation usage frames.
   */
  public completeStreaming(prompt: string): AsyncIterable<{ text: string }>;
  public completeStreaming(request: ChatRequest): AsyncIterable<{ text: string }>;
  public completeStreaming(input: string | ChatRequest): AsyncIterable<{ text: string }> {
    const request = typeof input === 'string' ? userTurnRequest(input) : input;
    const iter = this.completeAllDeltas(request);
    return {
      [Symbol.asyncIterator](): AsyncIterator<{ text: string }> {
        const inner = iter[Symbol.asyncIterator]();
        return {
          async next(): Promise<IteratorResult<{ text: string }>> {
            for (;;) {
              const step = await inner.next();
              if (step.done) return { value: undefined as unknown as { text: string }, done: true };
              if (step.value.kind === 'text') return { value: { text: step.value.text }, done: false };
              // Skip non-text deltas; the caller opted out of them by using this overload.
            }
          },
          async return(): Promise<IteratorResult<{ text: string }>> {
            if (inner.return) await inner.return(undefined);
            return { value: undefined as unknown as { text: string }, done: true };
          },
        };
      },
    };
  }

  /**
   * Streaming completion that surfaces every delta kind — text, reasoning,
   * tool calls, mid-generation usage frames and the terminal completion
   * event.
   */
  public completeAllDeltas(request: ChatRequest): AsyncIterable<Delta> {
    this.checkOpen();
    const subscriptionId = nextSubscriptionId('chat');
    return makeDeltaStream({
      subscriptionId,
      start: () =>
        NativeFoundryLocal.sessionCompleteStreaming(
          this.handle,
          JSON.stringify(encodeChatRequest(request)),
          subscriptionId,
        ),
    });
  }

  /**
   * Non-streaming completion. Returns the entire assistant response along
   * with any tool calls the model requested and, when the runtime reports
   * them, token usage counters.
   */
  public async complete(prompt: string): Promise<CompleteResult>;
  public async complete(request: ChatRequest): Promise<CompleteResult>;
  public async complete(input: string | ChatRequest): Promise<CompleteResult> {
    this.checkOpen();
    const request = typeof input === 'string' ? userTurnRequest(input) : input;
    try {
      const raw = await NativeFoundryLocal.sessionComplete(
        this.handle,
        JSON.stringify(encodeChatRequest(request)),
      );
      return parseCompleteResult(raw);
    } catch (err) {
      throw FoundryLocalError.fromNative(err);
    }
  }

  /**
   * Continue a turn by delivering results for tools the model asked to
   * call. Yields the next round of deltas — including the final assistant
   * text or a further round of tool calls.
   */
  public submitToolResults(results: readonly ToolResult[]): AsyncIterable<Delta> {
    this.checkOpen();
    const subscriptionId = nextSubscriptionId('tool');
    const payload = JSON.stringify(results.map((r) => ({ call_id: r.callId, result: r.resultJson })));
    return makeDeltaStream({
      subscriptionId,
      start: () =>
        NativeFoundryLocal.sessionSubmitToolResultsStreaming(this.handle, payload, subscriptionId),
    });
  }
}

// -----------------------------------------------------------------------------
// Audio
// -----------------------------------------------------------------------------

/** Speech-to-text session. */
export class AudioSession extends Session {
  /** @internal */
  public static create(model: Model, options: AudioOptions): AudioSession {
    const handle = NativeFoundryLocal.sessionCreate(
      model.nativeHandle,
      JSON.stringify({ type: 'audio', ...encodeAudioOptions(options) }),
    );
    return new AudioSession(handle);
  }

  private constructor(handle: number) {
    super(handle);
  }

  /** One-shot transcription. */
  public async transcribe(request: TranscribeRequest): Promise<TranscribeResult> {
    this.checkOpen();
    try {
      const raw = await NativeFoundryLocal.sessionTranscribe(
        this.handle,
        JSON.stringify(encodeTranscribeRequest(request)),
      );
      return parseTranscribeResult(raw);
    } catch (err) {
      throw FoundryLocalError.fromNative(err);
    }
  }

  /**
   * Streaming transcription. Emits `speechPartial` while listening and
   * `speechFinal` for stabilised segments; the terminal `completed` delta
   * carries the finish reason.
   */
  public transcribeStreaming(request: TranscribeRequest): AsyncIterable<Delta> {
    this.checkOpen();
    const subscriptionId = nextSubscriptionId('asr');
    return makeDeltaStream({
      subscriptionId,
      start: () =>
        NativeFoundryLocal.sessionTranscribeStreaming(
          this.handle,
          JSON.stringify(encodeTranscribeRequest(request)),
          subscriptionId,
        ),
    });
  }

  /**
   * Push a chunk of PCM audio into a live transcription session. Only
   * meaningful after `transcribeStreaming({ kind: 'streaming' })`.
   *
   * `pcm` is a `Uint8Array` of little-endian 16-bit PCM samples. It is
   * base64-encoded once on the JS side to cross the bridge.
   */
  public pushAudio(
    pcm: Uint8Array,
    sampleRate: number,
    channels: number,
    isFinal = false,
  ): void {
    this.checkOpen();
    NativeFoundryLocal.sessionPushAudio(
      this.handle,
      uint8ToBase64(pcm),
      sampleRate,
      channels,
      isFinal,
    );
  }
}

// -----------------------------------------------------------------------------
// Embedding
// -----------------------------------------------------------------------------

export class EmbeddingSession extends Session {
  /** @internal */
  public static create(model: Model, _options: EmbeddingOptions): EmbeddingSession {
    const handle = NativeFoundryLocal.sessionCreate(
      model.nativeHandle,
      JSON.stringify({ type: 'embedding' }),
    );
    return new EmbeddingSession(handle);
  }

  private constructor(handle: number) {
    super(handle);
  }

  public async embed(inputs: readonly string[]): Promise<EmbeddingResult> {
    this.checkOpen();
    try {
      const raw = await NativeFoundryLocal.sessionEmbed(
        this.handle,
        JSON.stringify({ inputs }),
      );
      return parseEmbeddingResult(raw);
    } catch (err) {
      throw FoundryLocalError.fromNative(err);
    }
  }
}

// -----------------------------------------------------------------------------
// Encoding / decoding helpers
// -----------------------------------------------------------------------------

function encodeChatOptions(options: ChatOptions): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (options.systemPrompt !== undefined) out.system_prompt = options.systemPrompt;
  if (options.temperature !== undefined) out.temperature = options.temperature;
  if (options.topP !== undefined) out.top_p = options.topP;
  if (options.topK !== undefined) out.top_k = options.topK;
  if (options.maxOutputTokens !== undefined) out.max_output_tokens = options.maxOutputTokens;
  if (options.seed !== undefined) out.seed = options.seed;
  out.keep_history = options.keepHistory ?? true;
  return out;
}

function encodeAudioOptions(options: AudioOptions): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (options.language !== undefined) out.language = options.language;
  if (options.maxOutputTokens !== undefined) out.max_output_tokens = options.maxOutputTokens;
  return out;
}

function encodeChatRequest(request: ChatRequest): Record<string, unknown> {
  const out: Record<string, unknown> = {
    messages: request.messages.map((m) => {
      const encoded: Record<string, unknown> = { role: m.role };
      if (typeof m.content === 'string') {
        encoded.content = m.content;
      } else {
        encoded.content = m.content.map((part) => {
          const p: Record<string, unknown> = { type: part.type };
          if ('text' in part && part.text !== undefined) p.text = part.text;
          if ('path' in part && part.path !== undefined) p.path = part.path;
          if ('dataBase64' in part && part.dataBase64 !== undefined) p.data_base64 = part.dataBase64;
          return p;
        });
      }
      if (m.toolCallId !== undefined) encoded.tool_call_id = m.toolCallId;
      return encoded;
    }),
  };
  if (request.tools && request.tools.length > 0) {
    out.tools = request.tools.map((t) => ({
      name: t.name,
      description: t.description,
      // parametersJson is a JSON Schema document; embed it as an object so
      // the ABI sees a schema, not a stringified schema.
      parameters: JSON.parse(t.parametersJson),
    }));
  }
  if (request.toolChoice !== undefined) out.tool_choice = request.toolChoice;
  if (request.temperature !== undefined) out.temperature = request.temperature;
  if (request.maxOutputTokens !== undefined) out.max_output_tokens = request.maxOutputTokens;
  return out;
}

function encodeTranscribeRequest(request: TranscribeRequest): Record<string, unknown> {
  switch (request.kind) {
    case 'file':
      return {
        path: request.path,
        ...(request.language !== undefined ? { language: request.language } : {}),
        translate: request.translate ?? false,
      };
    case 'inMemory':
      return {
        data_base64: request.dataBase64,
        format: request.format ?? 'pcm',
        sample_rate: request.sampleRate ?? 16000,
        channels: request.channels ?? 1,
      };
    case 'streaming':
      return {
        streaming: true,
        ...(request.language !== undefined ? { language: request.language } : {}),
      };
  }
}

function parseCompleteResult(raw: string | null | undefined): CompleteResult {
  const rawJson = raw ?? '';
  if (!rawJson) {
    return { text: '', finishReason: 'none', toolCalls: null, usage: null, rawJson: '' };
  }
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(rawJson) as Record<string, unknown>;
  } catch {
    return { text: '', finishReason: 'none', toolCalls: null, usage: null, rawJson };
  }

  const text = typeof obj.text === 'string' ? obj.text : '';
  const finishReason = parseFinishReason(
    typeof obj.finish_reason === 'string' ? (obj.finish_reason as string) : null,
  );

  // Absent vs empty: `null` when the core did not report tool_calls at all,
  // `[]` when it explicitly emitted an empty array. We keep the distinction.
  let toolCalls: readonly ToolCall[] | null = null;
  if (Array.isArray(obj.tool_calls)) {
    toolCalls = (obj.tool_calls as unknown[]).map((c) => {
      const call = c as { call_id?: unknown; name?: unknown; arguments?: unknown };
      return {
        callId: typeof call.call_id === 'string' ? call.call_id : '',
        name: typeof call.name === 'string' ? call.name : '',
        // Arguments stays a raw string; the model may emit something that
        // does not match the declared schema and it is the app's job to
        // decide what to do about that.
        argumentsJson: typeof call.arguments === 'string' ? call.arguments : '{}',
      };
    });
  }

  let usage: UsageStats | null = null;
  if (obj.usage && typeof obj.usage === 'object') {
    const u = obj.usage as { prompt_tokens?: unknown; completion_tokens?: unknown };
    usage = {
      promptTokens: toNumber(u.prompt_tokens),
      completionTokens: toNumber(u.completion_tokens),
    };
  }

  return { text, finishReason, toolCalls, usage, rawJson };
}

function parseTranscribeResult(raw: string | null | undefined): TranscribeResult {
  if (!raw) return { text: '', language: null, segments: [] };
  let obj: Record<string, unknown>;
  try { obj = JSON.parse(raw) as Record<string, unknown>; }
  catch { return { text: '', language: null, segments: [] }; }

  const segments = Array.isArray(obj.segments)
    ? (obj.segments as unknown[]).map((s) => {
        const seg = s as Record<string, unknown>;
        return {
          text: typeof seg.text === 'string' ? seg.text : '',
          startTimeMs: toNumber(seg.start_time_ms),
          endTimeMs: toNumber(seg.end_time_ms),
          language: typeof seg.language === 'string' ? seg.language : null,
        };
      })
    : [];

  return {
    text: typeof obj.text === 'string' ? obj.text : '',
    language: typeof obj.language === 'string' ? obj.language : null,
    segments,
  };
}

function parseEmbeddingResult(raw: string | null | undefined): EmbeddingResult {
  if (!raw) return { embeddings: [], dimensions: 0 };
  let obj: Record<string, unknown>;
  try { obj = JSON.parse(raw) as Record<string, unknown>; }
  catch { return { embeddings: [], dimensions: 0 }; }

  const dimensions = toNumber(obj.dimensions);
  const embeddings = Array.isArray(obj.embeddings)
    ? (obj.embeddings as unknown[]).map((row) =>
        Array.isArray(row) ? (row as unknown[]).map((n) => toNumber(n)) : [],
      )
    : [];
  return { embeddings, dimensions };
}

function userTurnRequest(prompt: string): ChatRequest {
  return { messages: [{ role: 'user', content: prompt }] };
}

// -----------------------------------------------------------------------------
// Small base64 helper for `pushAudio`. Kept local because RN environments
// vary in whether `Buffer`/`atob`/`btoa` is polyfilled.
// -----------------------------------------------------------------------------

const BASE64_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

function uint8ToBase64(bytes: Uint8Array): string {
  // Prefer the platform Buffer when Node.js is polyfilling it (Jest, Metro
  // dev server on some configurations); otherwise fall back to a manual
  // encoder so this file compiles standalone.
  const maybeBuffer = (globalThis as { Buffer?: { from: (b: Uint8Array) => { toString: (fmt: string) => string } } }).Buffer;
  if (maybeBuffer) return maybeBuffer.from(bytes).toString('base64');

  let out = '';
  let i = 0;
  for (; i + 2 < bytes.length; i += 3) {
    const b0 = bytes[i]!, b1 = bytes[i + 1]!, b2 = bytes[i + 2]!;
    out += BASE64_ALPHABET[b0 >> 2];
    out += BASE64_ALPHABET[((b0 & 0x03) << 4) | (b1 >> 4)];
    out += BASE64_ALPHABET[((b1 & 0x0f) << 2) | (b2 >> 6)];
    out += BASE64_ALPHABET[b2 & 0x3f];
  }
  const rem = bytes.length - i;
  if (rem === 1) {
    const b0 = bytes[i]!;
    out += BASE64_ALPHABET[b0 >> 2];
    out += BASE64_ALPHABET[(b0 & 0x03) << 4];
    out += '==';
  } else if (rem === 2) {
    const b0 = bytes[i]!, b1 = bytes[i + 1]!;
    out += BASE64_ALPHABET[b0 >> 2];
    out += BASE64_ALPHABET[((b0 & 0x03) << 4) | (b1 >> 4)];
    out += BASE64_ALPHABET[(b1 & 0x0f) << 2];
    out += '=';
  }
  return out;
}
